mne/time_frequency/psd.py

```diff
@@ -112,7 +113,220 @@ def _check_nfft(n, n_fft, n_overlap):
+                                                verbose=verbose)
+    x_splits = np.array_split(x, n_jobs)
+    f_psd = parallel(my_pwelch(d, noverlap=n_overlap, nfft=n_fft,
+                     fs=sfreq, freq_mask=freq_mask,
+                     welch_fun=welch)
+                     for d in x_splits)
+
+    # Combining/reshaping to original data shape
+    psds = np.concatenate(f_psd, axis=0)
+    psds = psds.reshape(np.hstack([dshape, -1]))
+    return psds, freqs
+
+
+@verbose
+def psd_welch(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None, n_fft=256,
+              n_overlap=0, picks=None, proj=False, n_jobs=1, verbose=None):
+    """Compute the power spectral density (PSD) using Welch's method.
+
+    Calculates periodigrams for a sliding window over the
+    time dimension, then averages them together for each channel/epoch.
+
+    Parameters
+    ----------
+    inst : instance of Epochs or Raw
+        The data for PSD calculation
+    fmin : float
+        Min frequency of interest
+    fmax : float
+        Max frequency of interest
+    tmin : float | None
+        Min time of interest
+    tmax : float | None
+        Max time of interest
+    n_fft : int
+        The length of the tapers ie. the windows. The smaller
+        it is the smoother are the PSDs. The default value is 256.
+        If ``n_fft > len(inst.times)``, it will be adjusted down to
+        ``len(inst.times)``.
+    n_overlap : int
+        The number of points of overlap between blocks. Will be adjusted
+        to be <= n_fft.
+        bandwidth : float
+        The bandwidth of the multi taper windowing function in Hz.
+    picks : array-like of int | None
+        The selection of channels to include in the computation.
+        If None, take all channels.
+    proj : bool
+        Apply SSP projection vectors. If inst is ndarray this is not used.
+    n_jobs : int
+        Number of CPUs to use in the computation.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    psds : ndarray, shape ([n_epochs], n_channels, n_freqs)
+        The power spectral densities. If Raw is provided,
+        then psds will be 2-D.
+    freqs : ndarray (n_freqs)
+        The frequencies.
+
+    See Also
+    --------
+    Raw.plot_psd, Epochs.plot_psd, psd_multitaper
+    """
+    # Prep data
+    data, sfreq = _check_psd_data(inst, tmin, tmax, picks, proj)
+    psds, freqs = _psd_welch(data, sfreq, fmin=fmin, fmax=fmax, n_fft=n_fft,
+                             n_overlap=n_overlap, n_jobs=n_jobs,
+                             verbose=verbose)
+    return psds, freqs
+
+
+def _psd_multitaper(x, sfreq, fmin=0, fmax=np.inf, bandwidth=None,
+                    adaptive=False, low_bias=True, normalization='length',
+                    n_jobs=1, verbose=None):
+    """Helper function for calculating Multitaper PSD."""
+    from .multitaper import multitaper_psd
```

Comment:
As in put the code for multitaper_psd into_psd_multitaper?
On Dec 28, 2015 2:10 PM, "Mainak Jas" <notifications@github.com> wrote:

> In mne/time_frequency/psd.py
> <https://github.com/mne-tools/mne-python/pull/2710#discussion_r48502634>:
>
> > +    --------
> > +    Raw.plot_psd, Epochs.plot_psd, psd_multitaper
> > +    """
> > +    # Prep data
> > +    data, sfreq = _check_psd_data(inst, tmin, tmax, picks, proj)
> > +    psds, freqs = _psd_welch(data, sfreq, fmin=fmin, fmax=fmax, n_fft=n_fft,
> > +                             n_overlap=n_overlap, n_jobs=n_jobs,
> > +                             verbose=verbose)
> > +    return psds, freqs
> > +
> > +
> > +def _psd_multitaper(x, sfreq, fmin=0, fmax=np.inf, bandwidth=None,
> > +                    adaptive=False, low_bias=True, normalization='length',
> > +                    n_jobs=1, verbose=None):
> > +    """Helper function for calculating Multitaper PSD."""
> > +    from .multitaper import multitaper_psd
>
> I just noticed this. I would actually do it the other way round. I would
> put the business logic in _psd_multitaper instead of multitaper_psd. That
> way, when you remove the function, you don't have to refactor again ... you
> just remove three lines of code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2710/files#r48502634>.
>